### PR TITLE
Support for manual CLIP loading in StableDiffusionPipeline - txt2img.

### LINKED ISSUE #3822: StableDiffusionPipeline() and CLIP "cooperation"
Hello!
First of all, amazing work, guys! Absolutely love the whole `transformers`, `diffusers` and others for its simplicity! ❤️ 

I would suggest an improvement to a `StableDiffusionPipeline()` function to allow for portable & offline execution on any machine if the required files are already in specified locations.
Right now, when running the following code, `StableDiffusionPipeline()`, it downloads an additional model: `openai/clip-vit-large-patch14`. 
```python
from diffusers import StableDiffusionPipeline
pipe = StableDiffusionPipeline.from_ckpt("path/to/single/safetensors/or/bin/file")
```
The problem is that the CLIP model downloads to cache, so it is impossible to create a truly portable & offline solution.
I suggest 2 solutions:
1. We load CLIP model ourselves and pass it as an argument to the `StableDiffusionPipeline`. If nothing is being passed, we go to current behavior. I definitely prefer this solution, as it does not break any previous code.
2. The `cache_dir` that we pass to `StableDiffusionPipeline` will also get passed to the CLIP model when it is called inside of `StableDiffusionPipeline`. This one, I am not so sure about, as there might be some edge cases, where this actually breaks smth.

### PR DESCRIPTION
**Pull request by WadRex**
# What does this PR do?
Fixes #3822 

This pull request introduces a feature that enhances the loading mechanism of the **`CLIP` model** when used in conjunction with `StableDiffusionPipeline.from_ckpt()`. This PR affects **only the `txt2img`** part of the mentioned pipeline. Users now have the flexibility to manually load the CLIP model and tokenizer, thereby bypassing the force loading behavior. This PR resolves the challenge of creating a fully portable solution, as the CLIP model and tokenizer would previously end up in the cache.
With this enhancement, users can now specify their desired CLIP model and tokenizer location as follows:
```python
# Users have the option to choose the official Hugging Face repository: `openai/clip-vit-large-patch14`,
# or provide the local path to load CLIP from.
clip_text_model = CLIPTextModel.from_pretrained("repo/id/or/path/to/local/folder")
clip_tokenizer = CLIPTokenizer.from_pretrained("repo/id/or/path/to/local/folder")

# Subsequently, these parameters are passed into the `StableDiffusionPipeline` function.
pipeline = StableDiffusionPipeline.from_ckpt("path/to/single/safetensors/or/bin/file", 
                                              clip_text_model=clip_text_model,
                                              clip_tokenizer=clip_tokenizer)
```
It is important to note that this pull request does not impact the behavior if users choose not to provide any parameters for `clip_text_model` or `clip_tokenizer`. In such cases, the code will behave as it did before the PR.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [X] Did you read the [contributor guideline](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md)?
- [x] Did you read our [philosophy doc](https://github.com/huggingface/diffusers/blob/main/PHILOSOPHY.md) (important for complex PRs)?
- [X] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link to it if that's the case. => Discussed in #3822 
- [X] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/diffusers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?
@patrickvonplaten 
@sayakpaul

### CODE REVIEW

File: src/diffusers/pipelines/stable_diffusion/convert_from_ckpt.py [hunk 1]
```diff
@@ -1022,7 +1022,9 @@ def download_from_original_stable_diffusion_ckpt(
     controlnet: Optional[bool] = None,
     load_safety_checker: bool = True,
     pipeline_class: DiffusionPipeline = None,
-    local_files_only=False,
+    local_files_only = False,
+    clip_text_model = None,
+    clip_tokenizer = None,
```
Comment (patrickvonplaten) [inline src/diffusers/pipelines/stable_diffusion/convert_from_ckpt.py:1027]:
```suggestion
    local_files_only=False,
    clip_text_model=None,
    clip_tokenizer=None,
```

**Review by patrickvonplaten — APPROVED** [review]
Looks great to me! @sayakpaul wdyt?

File: src/diffusers/loaders.py [hunk 2]
```diff
@@ -1339,6 +1339,10 @@ def from_ckpt(cls, pretrained_model_link_or_path, **kwargs):
                 "ddim"]`.
             load_safety_checker (`bool`, *optional*, defaults to `True`):
                 Whether to load the safety checker or not.
+            clip_text_model (`transformers.models.clip.modeling_clip.CLIPTextModel`, *optional*, defaults to `None`):
+                An instance of `CLIPTextModel` to use. If this parameter is `None`, the function will load a new instance of `CLIPTextModel`, if needed.
+            clip_tokenizer (`transformers.models.clip.tokenization_clip.CLIPTokenizer`, *optional*, defaults to `None`):
+                An instance of `CLIPTokenizer` to use. If this parameter is `None`, the function will load a new instance of `CLIPTokenizer`, if needed.
```
Comment (sayakpaul) [inline src/diffusers/loaders.py:1345]: Could we maybe follow something similar to for the docstrings here? 

https://github.com/huggingface/diffusers/blob/0bab447670f47c28df60fbd2f6a0f833f75a16f5/src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py#L90

File: src/diffusers/pipelines/stable_diffusion/convert_from_ckpt.py [hunk 3]
```diff
@@ -1071,6 +1073,10 @@ def download_from_original_stable_diffusion_ckpt(
             The pipeline class to use. Pass `None` to determine automatically.
         local_files_only (`bool`, *optional*, defaults to `False`):
             Whether or not to only look at local files (i.e., do not try to download the model).
+        clip_text_model (`transformers.models.clip.modeling_clip.CLIPTextModel`, *optional*, defaults to `None`):
+            An instance of `CLIPTextModel` to use. If this parameter is `None`, the function will load a new instance of `CLIPTextModel`, if needed.
+        clip_tokenizer (`transformers.models.clip.tokenization_clip.CLIPTokenizer`, *optional*, defaults to `None`):
+            An instance of `CLIPTokenizer` to use. If this parameter is `None`, the function will load a new instance of `CLIPTokenizer`, if needed.
```
Comment (sayakpaul) [inline src/diffusers/pipelines/stable_diffusion/convert_from_ckpt.py:1079]: Same as above.

**Review by sayakpaul — APPROVED** [review]
Thanks a lot for this important PR! Left some nits.

**Comment by sayakpaul** [comment]
Regarding the failing test [here](https://github.com/huggingface/diffusers/actions/runs/5338708493/jobs/9681359015?pr=3832), could you run `make style && make quality` from your environment? More info is available on: https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md

**Comment by WadRex** [comment]
@patrickvonplaten 
@sayakpaul 
Are there any more requirements or changes needed to successfully merge this PR?

**Pull request merged**